### PR TITLE
Fix Unix.getaddrinfo when called on strings containing null bytes

### PR DESCRIPTION
**Pull request by Armael**
```
utop # Unix.getaddrinfo "\000" "" [];;
Segmentation fault
```
(in batch mode, add a `Gc.full_major()` afterwards to crash the program)

This PR fixes a bug found and tracked down by Joe at the mirage hack retreat. `Unix.getaddrinfo` crashes the GC when called on a string containing null bytes because of a `return` that should be a `CAMLreturn`.

### CODE REVIEW

**Comment by shindere** [comment]
This looks very good to me, thakns.

I'd perhaps just reword the commit message and Changes entry because
"zero bytes" is unclear to me, I read it as being equivalent to an empty
string. Perhaps nul bytes, or just nuls or zeros, would be better?

**Review by shindere — APPROVED** [review]

**Comment by Armael** [comment]
Thanks for the review. I fixed the commit message as you said :).

**Comment by shindere** [comment]
Thanks! Let's wait for CI and I'll merge tomorrow if nobody has done it.

**Comment by gasche** [comment]
Are there other unprotected uses of `return` in caml-exposed C functions?

**Comment by shindere** [comment]
Gabriel Scherer (2018/10/10 12:56 -0700):
> Are there other unprotected uses of `return` in caml-exposed C
> functions?

I wondered, too. There are many uses of "return" in otherlibs/unix, but
there are so many of them that I thought they must be correct but that's
certainly something that would be worth investigating.

**Comment by Armael** [comment]
My understanding is that using `return` in bindings is fine as long as the function is noalloc and hence does not use `CAMLparam`/`CAMLlocal`. So just grepping for `return` is not enough; maybe this could be a nice opportunity to use Coccinelle to check for the correct use of these macros?

**Pull request merged**
